### PR TITLE
Adding ReCaptcha for Forgot Username and Password Scenario

### PR DESCRIPTION
**Pull request by sarubi**
### Proposed changes in this pull request

Add captcha validation for account recovery. i.e forgot username and password scenarios to the reduce brute force attacks.
This PR contains the fix for the above issue.
For more details,
https://docs.google.com/document/d/1ZDbpC7zhX-x_nKLNum2q0HZ_zD8FKVB7NAtOa5SNHLg/edit?usp=sharing

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+     */
+    public ReCaptchaProperties getReCaptcha(String tenantDomain, boolean isEndpointTenantAware, String captchaType,
+                                            String recoveryType) throws ApiException {
+
+        if (StringUtils.isBlank(tenantDomain)) {
+            tenantDomain = MultitenantConstants.SUPER_TENANT_DOMAIN_NAME;
+        }
+
+        if (isEndpointTenantAware && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equalsIgnoreCase(tenantDomain)) {
+            basePath = IdentityManagementServiceUtil.getInstance().getServiceContextURL()
```
Comment (GayanM) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java]: can't there be concurrent issue by modifying class variable?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        apiClient.setBasePath(basePath);
+
+        String localVarPath = "/captcha";
+
+        // Query params
+        List<Pair> localVarQueryParams = new ArrayList<>();
+        Map<String, String> localVarHeaderParams = new HashMap<>();
+        Map<String, Object> localVarFormParams = new HashMap<>();
+
+        localVarQueryParams.addAll(apiClient.parameterToPairs("", "tenant-domain", tenantDomain));
```
Comment (GayanM) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java]: can't use StringUtils.EMPTY ?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+        final String localVarAccept = apiClient.selectHeaderAccept(localVarAccepts);
+        final String localVarContentType = apiClient.selectHeaderContentType(localVarContentTypes);
+
+        String[] localVarAuthNames = new String[]{};
+
+        GenericType<ReCaptchaProperties> localVarReturnType = new GenericType<ReCaptchaProperties>() {
+        };
+
+        return apiClient.invokeAPI(localVarPath, "GET", localVarQueryParams, null, localVarHeaderParams,
```
Comment (GayanM) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java]: is there a http constant for "GET" ?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/SecurityQuestionApi.java [hunk 4]
```diff
@@ -96,6 +120,11 @@ public InitiateQuestionResponse securityQuestionGet(String username, String real
         // query params
         List<Pair> localVarQueryParams = new ArrayList<Pair>();
         Map<String, String> localVarHeaderParams = new HashMap<String, String>();
+
+        if (headers != null) {
```
Comment (GayanM) [inline components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/SecurityQuestionApi.java]: Can't use MapUtils.isNotEmpty ?

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java (hunk 2 above)
Comment (sarubi) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/ReCaptchaApi.java]: fixed

File: components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/SecurityQuestionApi.java (hunk 4 above)
Comment (sarubi) [reply components/identity-mgt/org.wso2.carbon.identity.mgt.endpoint/src/main/java/org/wso2/carbon/identity/mgt/endpoint/client/api/SecurityQuestionApi.java]: fixed

**Comment by sarubi** [comment]
Here I'm closing this PR and refer the link to improved version of this feature, [https://github.com/wso2/carbon-identity-framework/pull/1926](url)

**Pull request closed without merging**